File: WorldFeed/Server/src/WorldFeed.Common/Models/Properties/Geo.cs
﻿namespace WorldFeed.Common.Models.Properties
{
    using System.Collections.Generic;
    using System.Linq;
    using Newtonsoft.Json;

    using global::WorldFeed.Common.Extensions;
    using global::WorldFeed.Common.Public.Models;
    using global::WorldFeed.Common.Public.Models.Interfaces;

    /// <summary>
    /// Geographic information of a location
    /// </summary>
    public class Geo : IGeo
    {
        [JsonProperty("type")]
        public string Type { get; set; }

        // ReSharper disable once UnusedMember.Local -- This is used during the deserialization
        [JsonProperty("coordinates")]
        private List<double[][]> _storedCoordinates
        {
            set
            {
                if (value == null)
                {
                    Coordinates = null;
                }
                else if (value.IsEmpty())
                {
                    Coordinates = new List<ICoordinates>();
                }
                else
                {
                    var coordinatesInfo = value[0];
                    Coordinates = coordinatesInfo.Select(x => (ICoordinates)new Coordinates(x[1], x[0])).ToList();
                }
            }
        }

        [JsonIgnore]
        public List<ICoordinates> Coordinates { get; set; }
    }
}

File: Client/src/app/shared/posts/models/properties/geo.ts
import {IGeo} from "./IGeo";
import {ICoordinates} from "./ICoordinates";
import {Coordinates} from "./coordinates";

// Geographic information of a location
export class Geo implements IGeo {
    // [JsonProperty("type")]
    public type: string;

    // [JsonProperty("coordinates")]
    set _storedCoordinates(value: Array<number[][]>) {
        if (value === null) {
            this.coordinates = null;
        } else if (value.length <= 0) {
            this.coordinates = new Array<ICoordinates>();
        } else {
            let coordinatesInfo = value[0];
            this.coordinates = coordinatesInfo.map(x => new Coordinates(x[1], x[0]) as ICoordinates);
        }
    }

    // [JsonIgnore]
    public coordinates: Array<ICoordinates>;
}
